thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,219 @@
+        // NOTE - set skipLast to true when the oldest item must be skipped.  This should
+        // be used where the queue COULD have been updated after the buffers are swapped and
+        // the enqueueRowForTrace overwriting the oldest entry.
+        void queueOut(TraceQueue &traceOutQueue, bool skipOldestEntry)
+        {
+            const unsigned numberRowsToSkip = skipOldestEntry ? 1 : 0;
+            unsigned rowPos = rowTailp + numberRowsToSkip;
+            if (rowPos==rowBufferSize) rowPos = 0;
+            for (unsigned n=numberRowsToSkip; n<rowBufferSize;++n)
+            {
+                if (rowBuffer[rowPos])
+                {
+                    traceOutQueue.enqueue(rowBuffer[rowPos]);
+                    rowBuffer[rowPos].clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0;
+            }
+            traceOutQueue.setBatchCountFromSequence(getSequence());
+        }
+        // Dump items to MemoryBuffer (without clearing items from queue)
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+            IOutputMetaData *meta = helper->queryOutputMeta();
+            const __int64 cyclesNow = get_cycles_now();
+            const unsigned lastSequence = getSequence();
+            unsigned rowPos = rowTailp;
+            for (unsigned n=0; n<rowBufferSize;++n)
+            {
+                const CThorRowHistory &rowCurrent = rowBuffer[rowPos];
+                if (rowCurrent)
+                {
+                    xmlwrite.outputBeginNested("Row", true);
+                    const unsigned seq = lastSequence - (rowBufferSize - n);
+                    const __int64 age = cycle_to_millisec(cyclesNow-rowCurrent.getCycleStamp());
+                    xmlwrite.outputInt(seq, sizeof(seq), "@seq");
+                    xmlwrite.outputInt(meta->getRecordSize(rowCurrent.get()), sizeof(size32_t), "@size");
+                    xmlwrite.outputInt(age, sizeof(age), "@age");
+                    try
+                    {
+                        meta->toXML((const byte*)(rowCurrent.get()), xmlwrite);
+                    }
+                    catch (IException *e)
+                    {
+                        EXCLOG(e, "CTracingThorDataLink::dump");
+                        e->Release();
+                    }
+                    xmlwrite.outputEndNested("Row", true);
+                    mb.append(xmlwrite.str());
+                    xmlwrite.clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0;
+            }
+        }
+    };
+    const unsigned traceQueueSize;
+    class TraceQueue buffers[2];
+    class TraceQueue rowBufferLogCache;
+    atomic_t rowBufInUse;
+    Owned<IThorDataLink> thorDataLink;
+    IHThorArg *helper;
+
+    inline void enqueueRowForTrace(const void *row)
+    {
+        buffers[atomic_read(&rowBufInUse)].enqueue(row,queryEndCycles());
```

Comment:
minor: you could avoid a virtual call here, by calling thorDataLink->queryEndCycles() here directly, vs virtual queryEndCycles()